Link to license and changelog in overview

cuihtlauac opened the pull request:
Fix https://github.com/ocaml/ocaml.org/issues/36

Display links to changelog and license in overview, if available.

cuihtlauac committed:
Link to license, changelog and readme in overview

Fix https://github.com/ocaml/ocaml.org/issues/36

Display links to change log, readme, license in overview.
Also displays licence name.

cuihtlauac committed:
Do not generate broken changelog link

Skip changelog section if CHANGES.md file is missing

cuihtlauac committed:
Handle CHANGES.md and CHANGES

Some packages store their changelog in a file named CHANGES
instead of CHANGES.md. This allows to handle both cases.

tmattio commented on src/ocamlorg_frontend/pages/package_overview.eml:
```diff
@@ -80,33 +81,26 @@ Package_layout.render
             </a>
-            <a href="" class="flex items-center hover:text-primary-600">
+            <% | _ -> ()); %>
+            <a href="<%s Url.package_license package.name package.version %>" class="flex items-center hover:text-primary-600">
```
I think that won't work if the license file isn't `LICENSE.md`. For instance, if it's just `LICENSE`. We could pass a `license_file` field to the function like you do with `change_file`.

cuihtlauac replied:
Agreed, the `LICENSE` only file name (i.e. w/o .md) isn't properly handled.

tmattio commented on src/ocamlorg_package/lib/ocamlorg_package.ml:
```diff
@@ -301,15 +301,26 @@ let documentation_page ~kind t path =
+let readme_file ~kind = maybe_file "README.md.html" ~kind
 
-let license_file ~kind t =
+let license_file ~kind = maybe_file "LICENSE.md.html" ~kind
```
Related to the comment above: this could also be `LICENSE`.

tmattio commented on src/ocamlorg_package/lib/ocamlorg_package.ml:
```diff
@@ -301,15 +301,26 @@ let documentation_page ~kind t path =
+  match md with
+  | Some doc -> Lwt.return_some (true, doc)
+  | None ->
+    let+ txt = maybe_file "CHANGES.html" ~kind t in
```
Perhaps not useful if we only do this for the current function, but we could have a `maybe_files` function that takes a list of file names and will return the content of the first one that matches. So the `changes_files` would be implemented as:
```ocaml
let changes_file ~kind t =
  maybe_files [ "CHANGES.md.html"; "CHANGES.html"; "CHANGELOG.md.html" ]
```

tmattio commented on src/ocamlorg_package/lib/ocamlorg_package.mli:
```diff
@@ -104,6 +104,10 @@ val license_file :
 (** Get the license of a package *)
 
+val changes_file :
+  kind:[< `Package | `Universe of string ] -> t -> (bool * string) option Lwt.t
```
I'm not sure about returning a tuple with a bool that indicates whether this is a markdown file or not.

This is to be able to link to the correct file, but it could potentially be many formats. could we return the name of the file instead?

Also we should document what each of the tuple elements is.

tmattio commented on src/ocamlorg_web/lib/handler.ml:
```diff
@@ -414,6 +414,7 @@ let package_versioned t kind req =
         | None -> (description |> Omd.of_string |> Omd.to_html, "Description")
       in
+      let* changes_file = Ocamlorg_package.changes_file ~kind package in
       let _license = Ocamlorg_package.license_file ~kind package in
```
We're not passing the license file here, we could do that to check wether we were able to find it in the doc and link to it depending on that.

tmattio reviewed (commented):
Thanks a lot for the PR @cuihtlauac!

The approach looks great overall, I left a few comments with suggestions.

Commit:
Handle changelog and license file name variations

Both changelog and license file name candidates are searched:

* as plain text file (i.e. without extension)
* as markdown file (i.e. with .md extension).

Searched base name candidates are:

* License: LICENSE
* Changelog: CHANGES and CHANGELOG

In case other base names are missing, just add them to the list.

If either changelog or license is missing, no link will be
inserted in the overview.

tmattio committed:
Rename {license,changes}_file to {license,changes}_filename
